### PR TITLE
eth/tracers: add position field for callTracer logs

### PR DESCRIPTION
**Pull request by alvarosevilla95**
Currently, one can use the "withLog" parameter to include logs in the callTracer results, which allows the user to see at which trace level was each log emitted.

However, it does not permit to distinguish the exact ordering of a particular trace's logs and subexecutions. For example, if a trace emits a log, then calls another contract (subexecution starts), and after that returns it emits a second log, both would be included in the top trace's logs array, but one couldn't tell one was emitted before the call and the other after it.

This commit introduces a new field to the callLog struct named `Position`. It is equal to the current length of the calls array for the current trace at the time of the log emission. So a log that is emitted before any sub-calls are made will have position 0, and one that is emitted after a sub-call has happened will have position 1.

### CODE REVIEW

**Comment by alvarosevilla95** [comment]
@jsvisa tagging you as you developed the original feature. I found this added granularity useful for my use cases, thought I'd propose as a PR. I understand the added field may or may not be worth it for the majority of scenarios.

File: eth/tracers/native/call.go [hunk 1]
```diff
@@ -188,7 +189,9 @@ func (t *callTracer) CaptureState(pc uint64, op vm.OpCode, gas, cost uint64, sco
 			return
 		}
 
-		log := callLog{Address: scope.Contract.Address(), Topics: topics, Data: hexutil.Bytes(data)}
+		position := len(t.callstack[len(t.callstack)-1].Calls)
```
Comment (holiman) [inline eth/tracers/native/call.go:192]: Isn't `position` same thing as `depth` ?

**Comment by s1na** [comment]
> both would be included in the top trace's logs array, but one couldn't tell one was emitted before the call and the other after it.
Hm fair, I understand people care about this ordering.

How about using the `logIndex` field like we have them in `eth_getLogs` to stay more-or-less consistent? It will be a global incrementing index of logs.

**Comment by jsvisa** [comment]
> > both would be included in the top trace's logs array, but one couldn't tell one was emitted before the call and the other after it.
> > Hm fair, I understand people care about this ordering.
> 
> How about using the `logIndex` field like we have them in `eth_getLogs` to stay more-or-less consistent? It will be a global incrementing index of logs.

I've considered that option, too, but I think `index` is incremental and should not interrupt, but there may be an interruption in this use-case, such as:

```
Call
Log -> Position 0
Call
Call
Log -> Position 2
```

File: eth/tracers/native/call.go (hunk 1 above)
Comment (jsvisa) [reply eth/tracers/native/call.go:192]: I think they are not the same thing. here the log.position is used as relative to the number of calls in the current depth

**Comment by alvarosevilla95** [comment]
> > I've considered that option, too, but I think index is incremental and should not interrupt, but there may be an interruption in this use-case, such as:
> 
> I'm confused. In your case it's still possible to tell the ordering of logs by index.
> 
> ```
> Call
> Log -> Index 0
> Call
> Call
> Log -> Index 1
> ```

You would know the ordering of logs among themselves (but you already have this as the logs array is ordered). What I want to add here is a way to know the order of logs relative to calls made in the same trace. You want the second log to have a value of 2 so that you know that it was emitted after the second call.

As such I think using `index` for the field name could lead to confusion as the semantics are different than its use in `getLogs` (both in meaning and range). Admittedly, I'm not very happy with `Position` either, but I'm having trouble coming up with a better name that remains concise.

**Comment by alvarosevilla95** [comment]
> > > I've considered that option, too, but I think index is incremental and should not interrupt, but there may be an interruption in this use-case, such as:
> > 
> > 
> > I'm confused. In your case it's still possible to tell the ordering of logs by index.
> > ```
> > Call
> > Log -> Index 0
> > Call
> > Call
> > Log -> Index 1
> > ```
> 
> You would know the ordering of logs among themselves (but you already have this as the logs array is ordered). What I want to add here is a way to know the order of logs relative to calls made in the same trace. You want the second log to have a value of 2 so that you know that it was emitted after the second call.
> 
> As such I think using `index` for the field name could lead to confusion as the semantics are different from its use in `getLogs` (both in meaning and range). Admittedly, I'm not very happy with `Position` either, but I'm having trouble coming up with a better name that remains concise.

(for clarity, I'm assuming in the example above the first call is the one that initializes the trace. If the 3 calls listed were subcalls emitted by the trace, the log positions would be 1 and 3 respectively)

**Comment by alvarosevilla95** [comment]
Perhaps `CallOrder` would be more descriptive?

**Comment by s1na** [comment]
> What I want to add here is a way to know the order of logs relative to calls made in the same trace.

Ok I thought you want to have a global ordering of logs. Can you remind me please what is the use-case for this?

It is still possible to infer the information you want from a global index. Because there will be a gap in the index between for logs before and after a subcall. And you can compare the index of a log to that of a subcalls' logs to know after which subcall it exactly happened. Would this solve your problem? I want to know if we can re-use a field before we decide to add a new one. We should be careful with regards to reverts and fix up the indices (because logs of a reverted call are cleared). For `position` this is not an issue.

**Comment by alvarosevilla95** [comment]
> > What I want to add here is a way to know the order of logs relative to calls made in the same trace.
> 
> Ok I thought you want to have a global ordering of logs. Can you remind me please what is the use-case for this?
> 
> It is still possible to infer the information you want from a global index. Because there will be a gap in the index between for logs before and after a subcall. And you can compare the index of a log to that of a subcalls' logs to know after which subcall it exactly happened. Would this solve your problem? I want to know if we can re-use a field before we decide to add a new one. We should be careful with regards to reverts and fix up the indices (because logs of a reverted call are cleared). For `position` this is not an issue.

Ah, I think I see what you mean better now. Starting with you first question, the use case for the interleaved order for calls and logs for me is to display an "advanced trace" view, where every relevant action (call or log) is displayed, in the exact order that it happened. This can be useful when trying to understand the flow of execution of a transaction, where the exact order of calls made and logs emitted can help understand the logic executed.

Your proposed solution, if I understand correctly, is to reconstruct the global order of logs (such as you'd get from the tx receipt), in this trace structure. That would definitely be useful, as you don't need a second rpc call to be able to exactly sort the logs, but it would not entirely address my usecase. Think of the scenario where a subcall is made where no logs are emitted: the global index would remain unchanged, so you can't sort the logs wrt that subcall.

It is definitely a bit of a niche use case, and I see how for most people just a global log index, that allows to keep the original log order without the receipt, could make more sense. In that case I agree reusing the logIndex field would be best, as it would have the same purpose and values.

**Comment by jsvisa** [comment]
Here is a simple example to present what @alvarosevilla95 what to do.

```solidity
// SPDX-License-Identifier: GPL-3.0

pragma solidity >=0.8.2 <0.9.0;

contract Bar {
    uint x;
    function setX(uint _x) public {
        x = _x;
    }
}

contract Foo {
    event Call(address indexed from, address indexed to, uint indexed index);

    function callWithEvent(address _to) public {
        Bar bar = Bar(_to);
        bar.setX(0);
        emit Call(msg.sender, _to, 0);

        bar.setX(1);
        emit Call(msg.sender, _to, 1);
    }

    function callThenEvent(address _to) public {
        Bar bar = Bar(_to);
        bar.setX(0);
        bar.setX(1);
        emit Call(msg.sender, _to, 0);
        emit Call(msg.sender, _to, 1);
    }
}
```

> the trace for callThenEvent

```
debug.traceTransaction( '0x62025a8d5bee781f139034a85e89b7722f2e057e0e24e25b35619f907e5a5c10', {'tracer': 'callTracer', 'tracerConfig': {'withLog': true}})
```

```json
{
  "calls": [
    {
      "from": "0x52d3634b22b18b5076b696b9e597375b98bb3a33",
      "gas": "0x69bd",
      "gasUsed": "0xa26",
      "input": "0x4018d9aa0000000000000000000000000000000000000000000000000000000000000000",
      "to": "0xd79a837e2b3a2b370d14eec56850aaee7e83eb91",
      "type": "CALL",
      "value": "0x0"
    },
    {
      "from": "0x52d3634b22b18b5076b696b9e597375b98bb3a33",
      "gas": "0x5d63",
      "gasUsed": "0x4fae",
      "input": "0x4018d9aa0000000000000000000000000000000000000000000000000000000000000001",
      "to": "0xd79a837e2b3a2b370d14eec56850aaee7e83eb91",
      "type": "CALL",
      "value": "0x0"
    }
  ],
  "from": "0xf4d8223ff330debba976d219f459695090222640",
  "gas": "0xcce7",
  "gasUsed": "0xcce7",
  "input": "0x9864af23000000000000000000000000d79a837e2b3a2b370d14eec56850aaee7e83eb91",
  "logs": [
    {
      "address": "0x52d3634b22b18b5076b696b9e597375b98bb3a33",
      "data": "0x",
      "topics": [
        "0xda6f1030e46613a9add5390434c8e68a48f738007f0f5ab6986a52d0640a66f4",
        "0x000000000000000000000000f4d8223ff330debba976d219f459695090222640",
        "0x000000000000000000000000d79a837e2b3a2b370d14eec56850aaee7e83eb91",
        "0x0000000000000000000000000000000000000000000000000000000000000000"
      ]
    },
    {
      "address": "0x52d3634b22b18b5076b696b9e597375b98bb3a33",
      "data": "0x",
      "topics": [
        "0xda6f1030e46613a9add5390434c8e68a48f738007f0f5ab6986a52d0640a66f4",
        "0x000000000000000000000000f4d8223ff330debba976d219f459695090222640",
        "0x000000000000000000000000d79a837e2b3a2b370d14eec56850aaee7e83eb91",
        "0x0000000000000000000000000000000000000000000000000000000000000001"
      ]
    }
  ],
  "to": "0x52d3634b22b18b5076b696b9e597375b98bb3a33",
  "type": "CALL",
  "value": "0x0"
}
```

> traces for callWithEvent

```
debug.traceTransaction( '0x04e38e0a4033915e560741828699e86c104491efa52cf34c2dedbae90da2cd99', {'tracer': 'callTracer', 'tracerConfig': {'withLog': true}})
```

```json
{
  "calls": [
    {
      "from": "0x52d3634b22b18b5076b696b9e597375b98bb3a33",
      "gas": "0x2900",
      "gasUsed": "0x1516",
      "input": "0x4018d9aa0000000000000000000000000000000000000000000000000000000000000000",
      "to": "0xd79a837e2b3a2b370d14eec56850aaee7e83eb91",
      "type": "CALL",
      "value": "0x0"
    },
    {
      "from": "0x52d3634b22b18b5076b696b9e597375b98bb3a33",
      "gas": "0xa7e",
      "gasUsed": "0x1f2",
      "input": "0x4018d9aa0000000000000000000000000000000000000000000000000000000000000001",
      "to": "0xd79a837e2b3a2b370d14eec56850aaee7e83eb91",
      "type": "CALL",
      "value": "0x0"
    }
  ],
  "from": "0xf4d8223ff330debba976d219f459695090222640",
  "gas": "0x8b39",
  "gasUsed": "0x7f41",
  "input": "0xdccadbc2000000000000000000000000d79a837e2b3a2b370d14eec56850aaee7e83eb91",
  "logs": [
    {
      "address": "0x52d3634b22b18b5076b696b9e597375b98bb3a33",
      "data": "0x",
      "topics": [
        "0xda6f1030e46613a9add5390434c8e68a48f738007f0f5ab6986a52d0640a66f4",
        "0x000000000000000000000000f4d8223ff330debba976d219f459695090222640",
        "0x000000000000000000000000d79a837e2b3a2b370d14eec56850aaee7e83eb91",
        "0x0000000000000000000000000000000000000000000000000000000000000000"
      ]
    },
    {
      "address": "0x52d3634b22b18b5076b696b9e597375b98bb3a33",
      "data": "0x",
      "topics": [
        "0xda6f1030e46613a9add5390434c8e68a48f738007f0f5ab6986a52d0640a66f4",
        "0x000000000000000000000000f4d8223ff330debba976d219f459695090222640",
        "0x000000000000000000000000d79a837e2b3a2b370d14eec56850aaee7e83eb91",
        "0x0000000000000000000000000000000000000000000000000000000000000001"
      ]
    }
  ],
  "to": "0x52d3634b22b18b5076b696b9e597375b98bb3a33",
  "type": "CALL",
  "value": "0x0"
}
```

We can't find the difference between the two, and there is no way to know the order between the log and trace. But after this PR, we can see the differ

@alvarosevilla95 please correct me if I'm wrong.

**Comment by s1na** [comment]
> Think of the scenario where a subcall is made where no logs are emitted: the global index would remain unchanged, so you can't sort the logs wrt that subcall.

Yes your counter-example is valid. If there is no log emitted in a subcall it is not possible to say whether a log in the parent call was emitted before or after that subcall. I was mostly trying to understand if the index is enough, but seems like it's not. I'm fine with adding the position field.

@jsvisa thanks for the example!

File: eth/tracers/internal/tracetest/calltrace_test.go [hunk 2]
```diff
@@ -47,9 +47,10 @@ type callContext struct {
 
 // callLog is the result of LOG opCode
 type callLog struct {
-	Address common.Address `json:"address"`
-	Topics  []common.Hash  `json:"topics"`
-	Data    hexutil.Bytes  `json:"data"`
+	Address  common.Address `json:"address"`
+	Topics   []common.Hash  `json:"topics"`
+	Data     hexutil.Bytes  `json:"data"`
+	Position int            `json:"position"`
```
Comment (s1na) [inline eth/tracers/internal/tracetest/calltrace_test.go:53]: Please add a comment to indicate position is with respect to sub calls.

Comment (s1na) [reply eth/tracers/internal/tracetest/calltrace_test.go:53]: Sorry this is not the right place, I mean in the main tracer file not test file

Comment (alvarosevilla95) [reply eth/tracers/internal/tracetest/calltrace_test.go:53]: Added

**Comment by alvarosevilla95** [comment]
Yeah @jsvisa that's a perfect example, thanks!

**Review by s1na — APPROVED** [review]
SGTM

**Comment by s1na** [comment]
CI is failing, can you please fix?

```
>>> build/cache/golangci-lint-1.51.1-linux-amd64/golangci-lint run --config .golangci.yml ./...
eth/tracers/native/call.go:40: File is not `goimports`-ed (goimports)
	Address  common.Address `json:"address"`
	Topics   []common.Hash  `json:"topics"`
	Data     hexutil.Bytes  `json:"data"`
util.go:48: exit status 1
exit status 1
```

**Comment by alvarosevilla95** [comment]
> CI is failing, can you please fix?
> 
> ```
> >>> build/cache/golangci-lint-1.51.1-linux-amd64/golangci-lint run --config .golangci.yml ./...
> eth/tracers/native/call.go:40: File is not `goimports`-ed (goimports)
> 	Address  common.Address `json:"address"`
> 	Topics   []common.Hash  `json:"topics"`
> 	Data     hexutil.Bytes  `json:"data"`
> util.go:48: exit status 1
> exit status 1
> ```

Ah yes sorry about that. All green now

**Comment by jsvisa** [comment]
@s1na @alvarosevilla95 should we also add an `index` field to indicate the event's chronological order?

**Comment by s1na** [comment]
@jsvisa sure, but please make sure to have it consistent with the core log object. I believe it's called `logIndex`.

**Comment by jsvisa** [comment]
1. As we always return use hex-encoded integer instead of an integer, I changed the type of `Position` from `int` to `hexutil.Uint`. 
2. I also added a new `LogIndex` field. 
3. I added `WRITE_TEST_FILES` environment to rewrite the test response files, in this way you don't have to manually change a bunch of JSON files, so if we run the below command, it will rewrite the response file instead of the test checking:

```bash
WRITE_TEST_FILES=true go test ./eth/tracers/... -v
```

@s1na do you think this kind of adjustment is appropriate or not?

But seems I can't directly push to [alvarosevilla95:master](https://github.com/alvarosevilla95/go-ethereum/tree/master), so @alvarosevilla95 could you please cherry pick the commits from https://github.com/jsvisa/go-ethereum/tree/av-trace-position after @s1na approved those changes?

**Comment by s1na** [comment]
> As we always return use hex-encoded integer instead of an integer, I changed the type of Position from int to hexutil.Uint.

Ah good catch! glad we didn't merge this before this fix.

> I also added a new LogIndex field.

I checked your code and as I mentioned earlier it's not that simple. When there is a revert we will delete logs for some subcalls. That means those indices will be wiped. So all subsequent ones have to be updated.

**Comment by jsvisa** [comment]
> I checked your code and as I mentioned earlier it's not that simple. When there is a revert we will delete logs for some subcalls. That means those indices will be wiped. So all subsequent ones have to be updated.

Thanks for the tips, so let's move this one first(after fixing the hexadecimal), and implement the `logIndex` in another PR.

**Comment by alvarosevilla95** [comment]
Ah thanks for the catch about using hexutil.Uint @jsvisa, I missed the inconsistency there. Given the disccussion I've cherry-picked the position -> hex commits.

FWIW the write_test_files flags seems quite useful, would def have saved me time when changing the testcases, and in my case at least it was a relatively small change needed, but something bigger would get very annoying. But that can be decided on the follow up PR.

As a suggestion for logIndex, perhaps it makes sense to just add the logic to CaptureTxEnd, after the failed logs are cleared, and just assign all the indexes at that point.

**Comment by alvarosevilla95** [comment]
@s1na apologies but the CI is failing, but I can't spot what is it complaining about.. can you help spot it?

**Comment by s1na** [comment]
> @s1na apologies but the CI is failing and I can't spot what is it complaining about.. can you help identify it?

Ah don't mind that. That's just because the test suite times out on 32-bit build. It's a known issue.

> As a suggestion for logIndex, perhaps it makes sense to just add the logic to CaptureTxEnd, after the failed logs are cleared, and just assign all the indexes at that point.

I'm afraid it's not possible to determine the order at that point similar to the reason why it's not possible to determine the "position". In fact that's why I confused your PR in the first place :) But we'll figure it out in a future PR. This one is good to go!

**Review by s1na — APPROVED** [review]
LGTM

**Pull request merged**